bftengine/src/bftengine/KeyExchangeManager.cpp

```diff
@@ -131,6 +133,27 @@ void KeyExchangeManager::sendKeyExchange(const SeqNum& sn) {
   metrics_->sent_key_exchange_counter.Get().Inc();
 }
 
+// sends the clients public keys via the internal client, if keys weren't published or outdated.
+void KeyExchangeManager::sendInitialClientsKeys(const std::string& keys) {
+  if (clientsPublicKeys_.published_) {
+    LOG_INFO(ON_CHAIN_LOG, "Clients public keys were already published");
+    return;
+  }
+  auto ret = std::async(std::launch::async, [this, keys]() {
+    while (!initial_exchange_) {
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
+    auto cid = generateCid(kInitialClientsKeysCid);
+    LOG_DEBUG(ON_CHAIN_LOG, "Sends clients keys, cid: " << cid << " payload size " << keys.size());
+    client_->sendRequest(
+        bftEngine::CLIENTS_PUB_KEYS_FLAG | bftEngine::PUBLISH_ON_CHAIN_OBJECT_FLAG, keys.size(), keys.c_str(), cid);
+  });
+}
+
+bool KeyExchangeManager::isInitialClientKeysReq(const std::string& cid) {
+  return cid.find(kInitialClientsKeysCid) != std::string::npos;
```

Comment:
Can't we have another indication except the string comparison?

Reply:
the cid is the indication

Reply:
actually you're right, I have the validation flag on the request I'm using